iroh-net/src/endpoint.rs

```diff
@@ -0,0 +1,542 @@
+            quinn::EndpointConfig::default(),
+            server_config,
+            conn.clone(),
+            Arc::new(quinn::TokioRuntime),
+        )?;
+        trace!("created quinn endpoint");
+
+        Ok(Self {
+            keypair: Arc::new(keypair),
+            conn,
+            endpoint,
+            netmap: Arc::new(Mutex::new(netmap::NetworkMap { peers: vec![] })),
+            keylog,
+        })
+    }
+
+    /// Accept a connection on the socket.
+    pub fn accept(&self) -> quinn::Accept<'_> {
+        self.endpoint.accept()
+    }
+
+    /// Get the peer id of this endpoint.
+    pub fn peer_id(&self) -> PeerId {
+        self.keypair.public().into()
+    }
+
+    /// Get the local addresses on which the underlying magic socket is bound.
+    ///
+    /// Returns a tuple of the IPv4 and the optional IPv6 address.
+    pub fn local_addr(&self) -> anyhow::Result<(SocketAddr, Option<SocketAddr>)> {
+        self.conn.local_addr()
+    }
+
+    /// Get the local endpoints on which the underlying magic socket is listening.
+    ///
+    /// This returns information on the network state of the endpoints in addition to
+    /// their addresses.
+    pub async fn local_endpoints(&self) -> anyhow::Result<Vec<Endpoint>> {
+        self.conn.local_endpoints().await
+    }
+
+    /// Connect to a remote endpoint.
+    ///
+    /// The PeerId and the ALPN protocol are required. If you happen to know dialable addresses of
+    /// the remote endpoint, they can be specified and will be added to the endpoint's peer map.
+    /// If no addresses are specified, the endpoint will try to dial the peer through the
+    /// configured DERP servers.
+    pub async fn connect(
+        &self,
+        peer_id: PeerId,
+        alpn: &[u8],
+        known_addrs: &[SocketAddr],
+    ) -> anyhow::Result<quinn::Connection> {
+        self.add_known_addrs(peer_id, known_addrs).await?;
+
+        let node_key: hp::key::node::PublicKey = peer_id.into();
+        let addr = self
+            .conn
+            .get_mapping_addr(&node_key)
+            .await
+            .expect("just inserted");
+
+        let client_config = {
+            let alpn_protocols = vec![alpn.to_vec()];
+            let tls_client_config =
+                tls::make_client_config(&self.keypair, Some(peer_id), alpn_protocols, self.keylog)?;
+            let mut client_config = quinn::ClientConfig::new(Arc::new(tls_client_config));
+            let mut transport_config = quinn::TransportConfig::default();
+            transport_config.keep_alive_interval(Some(Duration::from_secs(1)));
+            client_config.transport_config(Arc::new(transport_config));
+            client_config
+        };
+
+        debug!(
+            "connecting to {}: (via {} - {:?})",
+            peer_id, addr, known_addrs
+        );
+        let connect = self
+            .endpoint
+            .connect_with(client_config, addr, "localhost")?;
```

Comment:
Can you add a todo, to replace `localhost` with sth that makes more sense? (this is old, and okay for now, but should be checked/fixed eventually)